### ISSUE TITLE
Modularization: kvstore

### ISSUE THREAD

**Issue by joamaki** [cilium/cilium#23732]
The kvstore package provides a client (BackendOperations) and a store abstraction (via JoinSharedStore)  used from the agent, operator and clustermesh-apiserver. These are used for reading and writing key-values described at https://docs.cilium.io/en/stable/kvstore/. 

The current implementation exposes the client via a public function kvstore.Client(), and the store via store.JoinSharedStore(). This makes integration testing (of dependees) and reuse of the kvstore harder.

The biggest blocker for modularization is the bootstrap problem. If the cluster is freshly initialized, we might not have kube-dns running yet and thus cannot perform name lookup for the kvstore hosts. The code works around this by using "ServiceCache" to figure out the IP address based on the K8s Service and Endpoints resources. Having the modularized kvstore depend on ServiceCache and K8sWatcher doesn't make sense so another solution is needed. We already have Resource[Service] and Resource[Endpoints] is implemented in `feature/service-refactor` branch and just needs to be ported to master. With these we can grab the synchronized stores and then lookup service and the associated endpoints to derive the backends.

Rough tasks:
- [ ] Create a cell for the client that provides the BackendOperations interface (rename type?)
- [ ] Solve the "CustomDialer" problem (above)
- [ ] Move Configuration options from DaemonConfig/OperatorConfig/etc. into kvstore.ClientConfig
- [ ] Move kvstore initialization in agent/operator/clustermesh into a kvstore start hook
- [ ] Remove kvstore.Client() function and wire the BackendOperations to use-sites
- [ ] Implement/adapt tests for the client cell

Nice to haves, but out of scope:
- Type safe interface to "SharedStore", e.g. "type Store[Obj Object] interface". This would clean up the confusion with "Key" being the object. Might make sense to use a workqueue to synchronize the local store rather than the full periodic iteration of all local key-value pairs.
- Event stream for store events, similar to Resource's Events(). We could perhaps find a common abstraction for Resource and kvstore?
- Global set of Stores, similar to pkg/k8s/shared_resources.go. Would need stores to be lazily initialized?

Related: #23425 
Blocked on: #23734

**Comment by joamaki** [comment]
ping @pippolo84

**Comment by joamaki** [comment]
Closing this as with the introduction of StateDB the approach described here no longer makes sense and this should be revisited.
